src/hotspot/share/gc/g1/g1BarrierSet.hpp

```diff
@@ -56,7 +56,9 @@ class G1BarrierSet: public CardTableBarrierSet {
   }
 
   // Add "pre_val" to a set of objects that may have been disconnected from the
-  // pre-marking object graph.
+  // pre-marking object graph. Prefer the version that takes location, as it
+  // can avoid touching the heap unnecessarily.
+  template <class T> static void enqueue_loc(T* dst);
```

Comment:
Since the "location" version should be preferred, I would have used

```cpp
templace <class T>
static void enqueue(T* dst);

static void enqueue_preloaded(oop pre_val);
```
to make the preferred version "default". Just my 2c.

Reply:
`enqueue_preloaded` it is! See new commit.